### PR TITLE
Allow filtering and sorting from synthetic non-Entity columns.

### PR DESCRIPTION
**Pull request by bobmcwhirter**
This is prereq for sorting by calculated columns like `average_score`

### CODE REVIEW

File: common/src/db/query.rs [hunk 1]
```diff
@@ -130,42 +145,210 @@ pub enum Error {
 // Internal types
 /////////////////////////////////////////////////////////////////////////
 
+#[derive(Default)]
+pub struct Columns {
+    columns: Vec<(ColumnRef, ColumnDef)>,
+}
```
Comment (jcrossley3) [inline common/src/db/query.rs:151]: If I'm understanding correctly, you're expanding the public interface of this module to include `Columns`, so the comment above it is wrong, since "internal types" aren't public.

File: common/src/db/query.rs [hunk 2]
```diff
@@ -130,42 +145,210 @@ pub enum Error {
...
-struct Filter<T: EntityTrait> {
-    operands: Operand<T>,
+struct Filter {
+    operands: Operand,
     operator: Operator,
 }
 
-struct Sort<T: EntityTrait> {
-    field: T::Column,
+impl Filter {
```
Comment (jcrossley3) [inline common/src/db/query.rs:200]: It's hard for me to see what changed here. I think it'd be easier to see if you edited the old code directly, instead of commenting it. (I kinda hate seeing huge commented-out sections of code anywhere)

File: common/src/db/query.rs [hunk 3]
```diff
@@ -188,6 +371,7 @@ impl<T: EntityTrait> From<Filter<T>> for ConditionExpression {
 // FromStr impls
 /////////////////////////////////////////////////////////////////////////
 
+/*
```
Comment (jcrossley3) [inline common/src/db/query.rs:374]: I know you like this technique. I don't. :)

File: common/src/db/query.rs [hunk 4]
```diff
@@ -420,18 +611,34 @@ mod tests {
...
-        assert!(Sort::<advisory::Entity>::from_str("location:desc").is_ok());
-        assert!(Sort::<advisory::Entity>::from_str("Location").is_ok());
-        assert!(Sort::<advisory::Entity>::from_str("Location:Asc").is_ok());
-        assert!(Sort::<advisory::Entity>::from_str("Location:Desc").is_ok());
+        assert!(Sort::from_str_with_entity::<advisory::Entity>("location").is_ok());
+        assert!(Sort::from_str_with_entity::<advisory::Entity>("location:asc").is_ok());
+        assert!(Sort::from_str_with_entity::<advisory::Entity>("location:desc").is_ok());
+        assert!(Sort::from_str_with_entity::<advisory::Entity>("Location").is_ok());
+        assert!(Sort::from_str_with_entity::<advisory::Entity>("Location:Asc").is_ok());
+        assert!(Sort::from_str_with_entity::<advisory::Entity>("Location:Desc").is_ok());
```
Comment (jcrossley3) [inline common/src/db/query.rs:619]: To me, the new name is awkward, perhaps because it's replacing a `FromStr` impl? I can't decide if introducing some new type, e.g. `SortedEntity` with its own `from_str` is an improvement or not.

**Review by jcrossley3 — CHANGES_REQUESTED** [review]
Clever idea, but I'm still fuzzy on how this will work with pagination. Won't this still be sorting each batch returned by the limiter instead of all the results matching the query?

Maybe a code snippet from the handler -- the caller of `filtering_with` -- interacting with the limiter would clear things up for me.

**Comment by bobmcwhirter** [comment]
Ultimately, all this is doing is allowing referencing in sort/filter to columns outside of the `E::Column` set.

We should be able to sort/filter for any column involved in the query, not just those tied to a given Entity.

I don't suppose it'll interact poorly with pagination, as these columns still poke right into the SQL.

e.g.

```
SELECT advisory.location, CHARLEN(advisory.location) as location_len FROM advisory ORDER BY location_len
```

And normal offset/limit can still apply.

File: common/src/db/query.rs [hunk 5]
```diff
@@ -126,77 +141,116 @@ pub enum Error {
...
+    }
 }
 
-struct Sort<T: EntityTrait> {
-    field: T::Column,
-    order: Order,
+impl IntoColumns for &Columns {
+    fn columns(&self) -> Columns {
+        (*self).clone()
+    }
```
Comment (jcrossley3) [inline common/src/db/query.rs:157]: Why is this necessary?

**Comment by jcrossley3** [comment]
@bobmcwhirter I have some concerns about excessive cloning. Because `Filter` and `Sort` aren't part of the public interface, I'm ok having them take a `&Columns` arg instead of an `IntoColumns` impl. Speaking of which, I think we can let that consume `self` to further simplify things.

I'm gonna approve this and push a subsequent PR along those lines.

**Review by jcrossley3 — APPROVED** [review]
